apps/aechannel/src/aesc_fsm.erl

```diff
@@ -2362,26 +2421,106 @@ send_conflict_msg(?WDRAW_ERR, Sn, Msg) ->
+                       [aesc_offchain_update:update()],
+                       aec_blocks:block_header_hash(),
+                       #data{}, send | defer) ->
+                              {ok, #data{}, [gen_statem:action()]}
+                            | {ok, fun(() -> ok), #data{}, [gen_statem:action()]}
+                            | {error, any()}.
+request_signing_(Tag, SignedTx, Updates, BlockHash, #data{client = Client} = D, SendAction) ->
     Info = #{signed_tx => SignedTx,
              updates => Updates},
     Msg = rpt_message(#{ type => sign
                        , tag  => Tag
                        , info => Info }, D),
-    Client ! {?MODULE, self(), Msg},
-    lager:debug("signing(~p) requested", [Tag]),
-    D#data{ op = #op_sign{ tag = Tag
-                         , data = #op_data{ signed_tx = SignedTx
-                                          , block_hash = BlockHash 
-                                          , updates = Updates}}
-          , log    = log_msg(req, sign, Msg, D#data.log)}.
+    D1 = D#data{ op = #op_sign{ tag = Tag
+                              , data = #op_data{ signed_tx = SignedTx
+                                               , block_hash = BlockHash 
+                                               , updates = Updates}}
+               , log    = log_msg(req, sign, Msg, D#data.log)},
+    SendF = sig_request_f(Client, Tag, Msg),
+    %% If we have a client pid: send the request
+    %% If not: proceed if our sig is already on the tx
+    %% Note that has_my_signature/3 is a bit costly, so we don't
+    %% check unless the client really is disconnected
+    IsOk = is_pid(Client) orelse
+        has_my_signature(Tag, SignedTx, D1),
+    case IsOk of
+        true ->
+            if is_pid(Client) ->
+                    req_signing_ok(SendAction, SendF, D1, []);
+               true ->
+                    req_signing_ok(SendAction, SendF, D1,
+                                   [{next_event, cast,
+                                     {?SIGNED, Tag, SignedTx}}])
+            end;
+        false ->
+            {error, client_disconnected}
+    end.
+
+%% When in a handle_call(), we want to reply to the caller before sending
+%% it some other message. In Erlang, using selective message reception, this
+%% is irrelevant, but in other environments (at the other end of a websocket)
+%% FIFO message handling can get confusing if we mix up the order. So in this
+%% case, pass the send fun and let the calling function decide.
+req_signing_ok(send, SendF, D, Actions) ->
+    SendF(),
+    {ok, D, Actions};
+req_signing_ok(defer, SendF, D, Actions) ->
+    {ok, SendF, D, Actions}.
+
+sig_request_f(undefined, _, _) ->
+    fun() -> ok end;
+sig_request_f(Client, Tag, Msg) when is_pid(Client) ->
+    fun() ->
+            Client ! {?MODULE, self(), Msg},
+            lager:debug("signing(~p) requested", [Tag])
+    end.
+
+has_my_signature(Tag, SignedTx, #data{ on_chain_id = ChannelPubkey } = D) ->
```

Comment:
Request signing is initiated only after proper sigs checks are done, hence this function my execute the very same checks for the second time. I am about to provide a small improvement that is based on the assumption that authentication is already checked

Reply:
Please consider incorporating #2628